File: androidApp/src/main/kotlin/com/torve/android/ui/device/DeviceLimitReachedScreen.kt
package com.torve.android.ui.device

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material.icons.filled.CheckCircle
import androidx.compose.material.icons.filled.Delete
import androidx.compose.material.icons.filled.DevicesOther
import androidx.compose.material.icons.filled.PhoneAndroid
import androidx.compose.material.icons.filled.Tablet
import androidx.compose.material.icons.filled.Tv
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.material3.TopAppBar
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import com.torve.android.ui.theme.Amber
import com.torve.android.ui.theme.Snow
import com.torve.data.device.ManagedDeviceDto
import com.torve.presentation.device.DeviceGovernanceViewModel
import org.koin.compose.koinInject

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun DeviceLimitReachedScreen(
    onBack: () -> Unit,
    onActivated: () -> Unit,
    viewModel: DeviceGovernanceViewModel = koinInject(),
) {
    val state by viewModel.state.collectAsState()

    LaunchedEffect(Unit) {
        viewModel.fetchDevices()
    }

    // Auto-navigate when current device becomes active
    LaunchedEffect(state.premiumAccess, state.activateSuccess) {
        if (state.premiumAccess || state.activateSuccess) {
            onActivated()
        }
    }

    Column(modifier = Modifier.fillMaxSize()) {
        TopAppBar(
            title = { Text("Device Limit Reached") },
            navigationIcon = {
                IconButton(onClick = onBack) {
                    Icon(Icons.AutoMirrored.Filled.ArrowBack, contentDescription = "Back")
                }
            },
        )

        LazyColumn(
            modifier = Modifier.fillMaxSize().padding(horizontal = 16.dp),
            verticalArrangement = Arrangement.spacedBy(8.dp),
        ) {
            item {
                Spacer(Modifier.height(8.dp))
                Icon(
                    Icons.Default.DevicesOther,
                    contentDescription = null,
                    tint = Amber,
                    modifier = Modifier.size(48.dp).fillMaxWidth(),
                )
                Spacer(Modifier.height(16.dp))
                Text(
                    text = "Your Torve Pro account is already active on ${state.maxActiveDevices} devices. Remove one device to activate Torve Pro on this device.",
                    style = MaterialTheme.typography.bodyLarge,
                    textAlign = TextAlign.Center,
                    modifier = Modifier.fillMaxWidth(),
                )
                Spacer(Modifier.height(24.dp))
                Text(
                    text = "Active Devices",
                    style = MaterialTheme.typography.titleMedium,
                    fontWeight = FontWeight.Bold,
                )
                Spacer(Modifier.height(8.dp))
            }

            items(
                state.devices.filter { it.is_active },
                key = { it.id },
            ) { device ->
                DeviceLimitRow(
                    device = device,
                    onRemove = { viewModel.removeDevice(device.id) },
                    isRemoving = state.isRemoving,
                )
            }

            state.error?.let { error ->
                item {
                    Spacer(Modifier.height(8.dp))
                    Text(
                        text = error,
                        color = MaterialTheme.colorScheme.error,
                        style = MaterialTheme.typography.bodySmall,
                    )
                }
            }

            item {
                Spacer(Modifier.height(16.dp))
                Text(
                    text = "To activate this device, remove one from your account.",
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                    textAlign = TextAlign.Center,
                    modifier = Modifier.fillMaxWidth(),
                )
                Spacer(Modifier.height(24.dp))
            }
        }
    }
}

@Composable
private fun DeviceLimitRow(
    device: ManagedDeviceDto,
    onRemove: () -> Unit,
    isRemoving: Boolean,
) {
    var showConfirm by remember { mutableStateOf(false) }
    val icon: ImageVector = when (device.device_type) {
        "tv" -> Icons.Default.Tv
        "tablet" -> Icons.Default.Tablet
        else -> Icons.Default.PhoneAndroid
    }

    Card(
        modifier = Modifier.fillMaxWidth(),
        colors = CardDefaults.cardColors(containerColor = MaterialTheme.colorScheme.surfaceVariant),
    ) {
        Row(
            modifier = Modifier.padding(16.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Icon(icon, contentDescription = null, tint = Amber, modifier = Modifier.size(28.dp))
            Spacer(Modifier.width(12.dp))
            Column(modifier = Modifier.weight(1f)) {
                Row(verticalAlignment = Alignment.CenterVertically) {
                    Text(
                        text = device.device_name,
                        style = MaterialTheme.typography.bodyLarge,
                        fontWeight = FontWeight.Medium,
                        color = Snow,
                    )
                    if (device.is_current) {
                        Spacer(Modifier.width(8.dp))
                        Text("THIS DEVICE", style = MaterialTheme.typography.labelSmall, color = Amber, fontWeight = FontWeight.Bold)
                    }
                }
                Text(
                    text = "${device.platform} · ${device.device_type}",
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )
            }

            if (!device.is_current) {
                IconButton(
                    onClick = { showConfirm = true },
                    enabled = !isRemoving,
                ) {
                    Icon(Icons.Default.Delete, contentDescription = "Remove", tint = MaterialTheme.colorScheme.error)
                }
            }
        }
    }

    if (showConfirm) {
        AlertDialog(
            onDismissRequest = { showConfirm = false },
            title = { Text("Remove Device") },
            text = { Text("Remove \"${device.device_name}\"? This will free a slot and Torve Pro will be activated on your current device.") },
            confirmButton = {
                TextButton(onClick = { showConfirm = false; onRemove() }) {
                    Text("Remove & Activate", color = MaterialTheme.colorScheme.error)
                }
            },
            dismissButton = {
                TextButton(onClick = { showConfirm = false }) { Text("Cancel") }
            },
        )
    }
}

File: iosApp/iosApp/Views/Settings/DeviceLimitReachedScreen.swift
import SwiftUI
import shared

struct DeviceLimitReachedScreen: View {
    @Environment(AppRouter.self) private var router
    @StateObject private var wrapper = DeviceGovernanceViewModelWrapper()
    @State private var deviceToRemove: ManagedDeviceDto? = nil

    var body: some View {
        List {
            headerSection

            Section("Active Devices") {
                ForEach(wrapper.state.devices.filter { ($0 as! ManagedDeviceDto).is_active }, id: \.id) { raw in
                    let device = raw as! ManagedDeviceDto
                    ActiveDeviceRow(device: device, onRemove: { deviceToRemove = device })
                }
            }

            Section {
                Text("To activate this device, remove one inactive or unused device from your account.")
                    .font(.caption)
                    .foregroundColor(SVColor.onSurfaceVariant)
            }
        }
        .navigationTitle("Device Limit Reached")
        .onAppear { wrapper.fetchDevices() }
        .onChange(of: wrapper.state.premiumAccess) { _, newValue in
            if newValue {
                router.popToRoot()
            }
        }
        .onChange(of: wrapper.state.activateSuccess) { _, newValue in
            if newValue {
                router.popToRoot()
            }
        }
        .confirmationDialog(
            "Remove Device",
            isPresented: Binding(
                get: { deviceToRemove != nil },
                set: { if !$0 { deviceToRemove = nil } }
            ),
            presenting: deviceToRemove
        ) { device in
            Button("Remove & Activate", role: .destructive) {
                wrapper.removeDevice(deviceId: device.id)
                deviceToRemove = nil
            }
            Button("Cancel", role: .cancel) { deviceToRemove = nil }
        } message: { device in
            Text("Remove \"\(device.device_name)\"? Your current device will be activated for this account.")
        }
    }

    private var headerSection: some View {
        Section {
            VStack(spacing: 16) {
                Image(systemName: "desktopcomputer.and.arrow.down")
                    .font(.system(size: 40))
                    .foregroundColor(SVColor.amber)

                Text(wrapper.state.deviceLimitKnown
                    ? "Your account is already active on \(wrapper.state.maxActiveDevices) devices. Remove one device to activate this device."
                    : "Checking your account device limit. Remove one active device to activate this device.")
                    .font(.subheadline)
                    .foregroundColor(SVColor.onSurfaceVariant)
                    .multilineTextAlignment(.center)

                if wrapper.state.activeDeviceCountKnown && wrapper.state.deviceLimitKnown {
                    Text("\(wrapper.state.activeDeviceCount) of \(wrapper.state.maxActiveDevices) devices active")
                        .font(.headline)
                        .foregroundColor(SVColor.amber)
                }
            }
            .frame(maxWidth: .infinity)
            .padding(.vertical, 8)
        }
    }
}

private struct ActiveDeviceRow: View {
    let device: ManagedDeviceDto
    let onRemove: () -> Void

    var body: some View {
        HStack(spacing: 12) {
            Image(systemName: iconName)
                .font(.title3)
                .foregroundColor(SVColor.amber)
                .frame(width: 32)

            VStack(alignment: .leading, spacing: 2) {
                HStack(spacing: 6) {
                    Text(device.device_name)
                        .fontWeight(.medium)
                    if device.is_current {
                        Text("THIS DEVICE")
                            .font(.caption2)
                            .fontWeight(.bold)
                            .foregroundColor(SVColor.amber)
                    }
                }
                Text("\(device.platform) · \(device.device_type)")
                    .font(.caption)
                    .foregroundColor(SVColor.onSurfaceVariant)
            }

            Spacer()

            if !device.is_current {
                Button(action: onRemove) {
                    Image(systemName: "trash")
                        .foregroundColor(SVColor.error)
                }
                .buttonStyle(.plain)
            }
        }
        .padding(.vertical, 4)
    }

    private var iconName: String {
        switch device.device_type {
        case "tv": return "tv"
        case "tablet": return "ipad"
        default: return "iphone"
        }
    }
}
